File: app/src/main/java/com/example/customviewcollection/draglayout/TestViewGroup.java
package com.example.customviewcollection.draglayout;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.core.view.ViewCompat;
import androidx.customview.widget.ViewDragHelper;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Locale;

public class TestViewGroup extends FrameLayout {

    private static final String TAG = "TestViewGroup";

    private ViewDragHelper mDragHelper;

    private boolean isMultiChoose = false;

    private ArrayList<View> viewList = new ArrayList<>();

    public boolean isMultiChoose() {
        return isMultiChoose;
    }

    public void setMultiChoose(boolean multiChoose) {
        isMultiChoose = multiChoose;
    }

    public TestViewGroup(Context context) {
        this(context,null,0);
    }

    public TestViewGroup(Context context, AttributeSet attrs) {
        this(context, attrs,0);
    }

    public TestViewGroup(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mDragHelper = ViewDragHelper.create(this, new ViewDragHelper.Callback() {
            @Override
            public boolean tryCaptureView(@NotNull View child, int pointerId) {
                return true;
            }

            @Override
            public int getViewHorizontalDragRange(@NotNull View child) {
//                if(mDragHelper.getCapturedView() == child) {
                   return getMeasuredWidth() - child.getMeasuredWidth();
//                }else {
//                    return 0;
//                }
            }

            public int getViewVerticalDragRange(@NotNull View child) {
                return getMeasuredHeight() - child.getMeasuredHeight();
            }

            @Override
            public void onViewDragStateChanged(int state) {
                super.onViewDragStateChanged(state);
                Log.e(TAG,"onViewDragStateChanged STATE: " + state);
                if(state == ViewDragHelper.STATE_SETTLING) {
                    mDragHelper.abort();
                }
            }

            @Override
            public void onViewPositionChanged(@NonNull View changedView, int left, int top, int dx, int dy) {
                super.onViewPositionChanged(changedView, left, top, dx, dy);
                Log.e(TAG, "onViewPositionChanged " + "left:" + left + ",top:" + top + ",dx:" + dx + ",dy:" + dy );
//                invalidate();

            }

            @Override
            public int clampViewPositionHorizontal(@NotNull View child, int left, int dx) {
                if (left < 0) {
                    left = 0;
                } else if (left > (getMeasuredWidth() - child.getMeasuredWidth())) {
                    left = getMeasuredWidth() - child.getMeasuredWidth();
                }
                return left;
            }

            public int clampViewPositionVertical(@NotNull View child, int top, int dy) {
                if (top < 0) {
                    top = 0;
                } else if (top > getMeasuredHeight() - child.getMeasuredHeight()) {
                    top = getMeasuredHeight() - child.getMeasuredHeight();
                }
                return top;
            }

            @Override
            public void onViewReleased(@NotNull View releasedChild, float xvel, float yvel) {
                int centerLeft = getMeasuredWidth() / 2 - releasedChild.getMeasuredWidth() / 2;
                if (releasedChild.getLeft() < centerLeft) {
//                    mDragHelper.settleCapturedViewAt(0, releasedChild.getTop());
//                    ViewCompat.postInvalidateOnAnimation(TestViewGroup.this);
                    mDragHelper.smoothSlideViewTo(releasedChild, 0, releasedChild.getTop());
                } else {
//                    mDragHelper.settleCapturedViewAt(getMeasuredWidth() - releasedChild.getMeasuredWidth(), releasedChild.getTop());
                    mDragHelper.smoothSlideViewTo(releasedChild, getMeasuredWidth() - releasedChild.getMeasuredWidth(), releasedChild.getTop());
//                    ViewCompat.postInvalidateOnAnimation(TestViewGroup.this);
                }
                ViewCompat.postInvalidateOnAnimation(TestViewGroup.this);
//                this.onViewDragStateChanged(ViewDragHelper.STATE_SETTLING);
//                mDragHelper.settleCapturedViewAt(mDragOriLeft, mDragOriTop);
//                invalidate();
            }


        });


    }

    private float startX;
//    private long downTime;

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        boolean result = mDragHelper.shouldInterceptTouchEvent(ev);
        float curX = ev.getX();
        switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
                startX = curX;
//                downTime = ev.getDownTime();
                break;
            case MotionEvent.ACTION_MOVE:
                Log.e(TAG,"onInterceptTouchEvent space: " + String.format(Locale.CHINA,"%.2f",curX - startX));
                if(Math.abs(curX - startX) > mDragHelper.getTouchSlop()) result = true;
                break;
//            case MotionEvent.ACTION_UP:
//                long time = ev.getEventTime();
//                if(time - downTime > 1000) {
//                    result = true;
//                }
//                break;
        }
        Log.e(TAG,"onInterceptTouchEvent result: " + result);
        return result;
    }

    @Override
    public boolean performClick() {
        return super.performClick();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mDragHelper.processTouchEvent(event);
        return true;
    }

    int mVdhXOffset;
    int mVdhYOffset;

    @Override
    public void computeScroll() {
        super.computeScroll();
        if(mDragHelper != null) {
            if(mDragHelper.continueSettling(true)) {
                ViewCompat.postInvalidateOnAnimation(TestViewGroup.this);
            }else {
                if(mDragHelper.getCapturedView() != null) {
                    mVdhXOffset = mDragHelper.getCapturedView().getLeft();
                    mVdhYOffset = mDragHelper.getCapturedView().getTop();
                }
            }
        }
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if(null != mDragHelper) {
            if(mDragHelper.getCapturedView() != null) {
                Log.e(TAG,"onLayout");
                mDragHelper.getCapturedView().offsetLeftAndRight(mVdhXOffset);
                mDragHelper.getCapturedView().offsetTopAndBottom(mVdhYOffset);
            }
        }
    }
}

File: app/src/main/java/com/example/customviewcollection/draglayout/TestViewGroup.kt
package com.example.customviewcollection.draglayout

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import android.widget.Toast
import androidx.core.view.ViewCompat
import androidx.customview.widget.ViewDragHelper
import java.util.*
import kotlin.math.abs

class TestViewGroup @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {

    private lateinit var mDragHelper: ViewDragHelper

    var list = ArrayList<Int>()
    var isMultiChoose = false

    private var startX = 0f

    //    private long downTime;
    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        var result = mDragHelper.shouldInterceptTouchEvent(ev)
        val curX = ev.x
        when (ev.action) {
            MotionEvent.ACTION_DOWN -> startX = curX
            MotionEvent.ACTION_MOVE -> {
                Log.e(TAG, "onInterceptTouchEvent space: " + String.format(Locale.CHINA, "%.2f", curX - startX))
                if (abs(curX - startX) > mDragHelper.touchSlop) result = true
            }
        }
        Log.e(TAG, "onInterceptTouchEvent result: $result")
        return result
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        for (i in 0 until childCount) {
            getChildAt(i).tag = i.toString()
            getChildAt(i).setOnClickListener { v: View ->
                if (isMultiChoose) {
                    list.add(v.tag.toString().toInt())
                } else {
                    Toast.makeText(this.context, "旋转", Toast.LENGTH_SHORT).show()
                    v.rotation = Random().nextFloat()
                }
            }
        }
    }

    override fun performClick(): Boolean {
        Log.e(TAG,"performClick")
        return super.performClick()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        performClick()
        mDragHelper.processTouchEvent(event)
        return true
    }

    companion object {
        private const val TAG = "TestViewGroup"
    }

    init {
        mDragHelper = ViewDragHelper.create(this, object : ViewDragHelper.Callback() {
            override fun tryCaptureView(child: View, pointerId: Int): Boolean { return true }

            override fun getViewHorizontalDragRange(child: View): Int {
                return measuredWidth - child.measuredWidth
            }

            override fun getViewVerticalDragRange(child: View): Int {
                return measuredHeight - child.measuredHeight
            }

            override fun onViewDragStateChanged(state: Int) {
                super.onViewDragStateChanged(state)
                Log.e(TAG, "onViewDragStateChanged STATE: $state")
                if (state == ViewDragHelper.STATE_SETTLING) {
                    mDragHelper.abort()
                }
            }

            override fun onViewPositionChanged(changedView: View, left: Int, top: Int, dx: Int, dy: Int) {
                super.onViewPositionChanged(changedView, left, top, dx, dy)
                Log.e(TAG, "onViewPositionChanged " + "left:" + left + "top:" + top + ",dx:" + dx + ",dy:" + dy)
            }

            override fun clampViewPositionHorizontal(child: View, left: Int, dx: Int): Int {
                var left1 = left
                if (left1 < 0) {
                    left1 = 0
                } else if (left1 > measuredWidth - child.measuredWidth) {
                    left1 = measuredWidth - child.measuredWidth
                }
                return left1
            }

            override fun clampViewPositionVertical(child: View, top: Int, dy: Int): Int {
                var top1 = top
                if (top1 < 0) {
                    top1 = 0
                } else if (top1 > measuredHeight - child.measuredHeight) {
                    top1 = measuredHeight - child.measuredHeight
                }
                return top1
            }

            override fun onViewReleased(releasedChild: View, xvel: Float, yvel: Float) {
                val centerLeft = measuredWidth / 2 - releasedChild.measuredWidth / 2
                if (releasedChild.left < centerLeft) {
                    mDragHelper.smoothSlideViewTo(releasedChild, 0, releasedChild.top)
                } else {
                    mDragHelper.smoothSlideViewTo(releasedChild, measuredWidth - releasedChild.measuredWidth, releasedChild.top)
                }
                Log.e(TAG, "onViewReleased " + "left" + releasedChild.left + "top" + releasedChild.top + "right" + releasedChild.right + "bottom" + releasedChild.bottom)
                Log.e(TAG, "onViewReleased raction " + releasedChild.rotation)
                ViewCompat.postInvalidateOnAnimation(this@TestViewGroup)

            }
        })
    }
}
